src/s3ql/backends/gdrive.py

```diff
@@ -0,0 +1,599 @@
+            else:
+                properties[k] = v
+        return properties
+
+    def _decode_metadata(self, f):
+        metadata = dict()
+        properties = f.get('properties', dict())
+        for k, v in properties.items():
+            if k.startswith('ref('):
+                continue
+            if v.startswith(Backend.VALUE_BASE64):
+                metadata[k] = base64.b64decode(v[len(Backend.VALUE_BASE64):])
+            elif v.startswith(Backend.VALUE_REF):
+                b64 = ''
+                for i in range(int(v[len(Backend.VALUE_REF):])):
+                    b64 += properties[self._chunk_property(k, i)]
+                metadata[k] = base64.b64decode(b64)
+            elif v.isdigit():
+                metadata[k] = int(v)
+            else:
+                metadata[k] = v
+        log.debug('{0}: metadata {1}'.format(f['name'], metadata))
+        return metadata
+
+    @property
+    @copy_ancestor_docstring
+    def has_native_rename(self):
+        return True
+
+    def __str__(self):
+        return 'gdrive folder %s@%s' % (self.login, self.prefix)
+
+    @retry
+    @copy_ancestor_docstring
+    def lookup(self, key):
+        log.debug("key: {0}".format(key))
+        f = self._lookup_file(key, self.folder)
+        if f is None:
+            raise NoSuchObject(key)    
+        return self._decode_metadata(f)
+
+    @retry
+    @copy_ancestor_docstring
+    def get_size(self, key):
+        log.debug("key: {0}".format(key))
+        f = self._lookup_file(key, self.folder)
+        if f is None:
+            raise NoSuchObject(key)
+        return int(f.get('size', u'0'))
+    @retry
+    @copy_ancestor_docstring
+    def open_read(self, key):
+        log.debug("key: {0}".format(key))
+        f = self._lookup_file(key, self.folder)
+        if f is None:
+            raise NoSuchObject(key)  
+        return ObjectR(self.service, f, self._decode_metadata(f))
+
+    @prepend_ancestor_docstring
+    def open_write(self, key, metadata=None, is_compressed=False):
+        log.debug("key: {0}".format(key))
+        if metadata is None:
+            metadata = dict()
+        elif not isinstance(metadata, dict):
+            raise TypeError('*metadata*: expected dict or None, got %s' % type(metadata))
+
+        f = self._lookup_file(key, self.folder)
+        if f is not None:
+            log.debug("Object {0} already exist!".format(key))
+            self._delete_by_id(f['id'])
+        body = {
+            'name': key,
+            'mimeType': Backend.MIME_TYPE_BINARY,
+            'parents': [ self.folder['id'] ],
+            'properties': self._encode_metadata(metadata),
+        }
+        log.debug("metadata: {0}, body: {1}".format(metadata, body))
+        return ObjectW(self.service,self, key,body)
+
+    @prepend_ancestor_docstring
+    def clear(self):
+        log.debug("")
+        for f in self._list_files(self.folder, "id"):
+            self._delete_by_id(f['id'])            
+    '''
```

Comment:
What's happening here? Why is this all commented out?

Reply:
it was an attemt to implement delete_multi, but it is very hard because gdrive no process all of them as block, most of them could fail, what do you do in this case? rexecute until 0 pending/error requests? probably you will get a ban for too much requests.

Reply:
I do not understand what you mean with *because gdrive no process all of them as block, most of them could fail*. If there is no way to implement `delete_multi`, just don't implement it. The higher layers will fail back to repeated `delete` calls instead.

I expect that the server will send you warnings to slow down before banning you completely. If the method raises the right exceptions, S3QL will honor these warnings and everything should work fine. If the *google-api-python-client* module doesn't give you access to these early warnings then... we shouldn't use it.